pycue/opencue/cloud/api.py

```diff
@@ -0,0 +1,144 @@
+#  Copyright Contributors to the OpenCue Project
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+
+
+class CloudInstanceGroup(object):
+    def __init__(self, data):
+        self.data = data
+        self.instances = []
+
+    @abc.abstractmethod
+    def get_instances(self):
+        """
+        Gets all instances of the current group
+        :return: (list) List of instances
+        """
+
+    @abc.abstractmethod
+    def resize(self, size=None):
+        """
+        Resizes the group to the given number of instances
+        :param size: (int) The target size of group to be scaled up/down to
+        :return:
+        """
+
+    @abc.abstractmethod
+    def name(self):
+        """
+        Name of the cloud group
+        :return:
+        """
+
+    @abc.abstractmethod
+    def status(self):
+        """
+        Returns the status of any operation made on the group. By default "STABLE"
+        Each provider has it's own way of implementing status of a particular group
+        :return:
+        """
+
+    @abc.abstractmethod
+    def id(self):
+        """
+        Used to treat a CloudInstanceGroup object similar to a gRPC object for threadpool usage
+        :return:
+        """
+
+    @abc.abstractmethod
+    def delete_cloud_group(self):
+        """
+        Delete the cloud group from the provider
+        :return:
+        """
+
+    @abc.abstractmethod
+    def current_group_size_info(self):
+        """
+        Implement this method to show the state of the instances for the group
+        If resizing ongoing: current_size -> target_size
+        If not: current_size (or) just the len(instances)
+        :return:
+        """
+
+
+class CloudManager(object):
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def get_registered_providers():
+        """
+        :return:
```

Comment:
Make sure to go through and document all the methods here. What's returned here is particularly important.